web-registry/src/features/marketplace/BuySellOrderFlow/BuySellOrderFlow.tsx

```diff
@@ -107,23 +97,95 @@ export const BuySellOrderFlow = ({
     error,
     setError,
   } = useMsgClient(handleTxQueued, handleTxDelivered, handleError);
+
   const accountAddress = wallet?.address;
   const txHash = deliverTxResponse?.transactionHash;
   const txHashUrl = getHashUrl(txHash);
   const errorEnum = findErrorByCodeEnum({ errorCode: error });
   const ErrorIcon = errorsMapping[errorEnum].icon;
 
-  const project = useMemo(
-    () => ({
-      id: selectedProject?.id.toString() ?? '',
-      sellOrders: projectUiSellOrdersInfo?.filter(
+  /**
+   * data processing (business logic)
+   */
+  const [selectedProject, setSelectedProject] =
+    useState<ProjectWithOrderData>();
+
+  useEffect(() => {
+    if (projects?.length === 1) setSelectedProject(projects[0]);
+  }, [projects]);
+
+  const setSelectedProjectById = (projectId: string): void | undefined => {
+    if (!projects || projects.length <= 1) return;
+    const found = projects.find(project => project.id === projectId);
+    if (found) setSelectedProject(found);
+  };
+
+  const sellOrderIds = useMemo(
+    () =>
+      projects &&
+      projects
+        .map(project => project.sellOrders)
+        .flat()
+        .map(sellOrder => sellOrder.id),
+    [projects],
+  );
+
+  const { sellOrdersResponse, refetchSellOrders } = useQuerySellOrders();
+  const allSellOrders = sellOrdersResponse?.sellOrders;
+
+  const projectUiSellOrdersInfo = useMemo(
+    () =>
+      allSellOrders
+        ?.map(normalizeToUISellOrderInfo)
+        .filter(sellOrder => sellOrderIds?.includes(sellOrder.id)),
+    [allSellOrders, sellOrderIds],
+  );
+
+  // project data prepared for buy modal (add projectId and filter the seller)
+  const _project = useMemo(
+    () =>
+      selectedProject && {
+        id: selectedProject?.id.toString() ?? '',
+        sellOrders: projectUiSellOrdersInfo?.filter(
+          sellOrder => sellOrder.seller !== accountAddress,
+        ),
+      },
+    [selectedProject, projectUiSellOrdersInfo, accountAddress],
+  );
+
+  const _sellOrders = useMemo(
+    () =>
+      projectUiSellOrdersInfo?.filter(
         sellOrder => sellOrder.seller !== accountAddress,
```

Comment:
Seems like we're doing that filtering twice now. Should we put that variable above and use it in `_project`?

Reply:
@flagrede Along with the above comments I have refactored this part accordingly. Nice!